From 3215cef562956dfdad2464e83c8d4c7e906da1bb Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:49:03 -0400
Subject: [tools] perf: Fine tune readn function

Message-id: <1408953210-26343-56-git-send-email-jolsa@redhat.com>
Patchwork-id: 88762
O-Subject: [PATCH RHEL7.1 BZ1133083 055/322] perf tools: Fine tune readn function
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 838d14520267769648fb2fc2a637107a1d102590
Author: Jiri Olsa <jolsa@redhat.com>
Date: Thu Nov 28 11:30:15 2013 +0100

description
===========
Added a 'left' variable to make the flow clearer, and added a debug
check for the return value - returning 'n' is more obvious.

Added small comment for readn.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/util/util.c b/tools/perf/util/util.c
index 9440481..6ea0b4a 100644
--- a/tools/perf/util/util.c
+++ b/tools/perf/util/util.c
@@ -6,6 +6,7 @@
 #endif
 #include <stdio.h>
 #include <stdlib.h>
+#include <linux/kernel.h>
 
 /*
  * XXX We need to find a better place for these things...
@@ -151,21 +152,26 @@ unsigned long convert_unit(unsigned long value, char *unit)
  return value;
 }
 
+/*
+ * Read exactly 'n' bytes or return an error.
+ */
 ssize_t readn(int fd, void *buf, size_t n)
 {
  void *buf_start = buf;
+ size_t left = n;
 
- while (n) {
-  ssize_t ret = read(fd, buf, n);
+ while (left) {
+  ssize_t ret = read(fd, buf, left);
 
   if (ret <= 0)
    return ret;
 
-  n -= ret;
-  buf += ret;
+  left -= ret;
+  buf  += ret;
  }
 
- return buf - buf_start;
+ BUG_ON((size_t)(buf - buf_start) != n);
+ return n;
 }
 
 size_t hex_width(u64 v)
-- 
1.7.1